service/storage_service.cc

```diff
@@ -5189,6 +5189,17 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                 // we already did read barrier above
                 result.status = raft_topology_cmd_result::command_status::success;
             break;
+            case raft_topology_cmd::command::barrier_normal:
+                // we already did the barrier, but we need to check
+                // whether we are in normal mode
+                if (_operation_mode != mode::NORMAL) {
+                    co_await coroutine::return_exception(std::runtime_error(
```

Comment:
We discussed offline that the `barrier_normal` rpc will be changed to a barrier that succeeds only after the node has updated its features via in `system.topology` after start. After this change, I can safely say that the exception should be pretty rare.

Reply:
How can it ever happen? The coordinator suppose to issue the barrier only when all nodes are in normal state and there are features to be enabled. It should be an assert.

Reply:
It's a misuse of the `_operation_mode` from my side. I need to give a guarantee to the topology coordinator here that, after responding, this node won't update its `supported_features` column without doing the feature check. It can happen if a node handles a barrier request early during boot before calling `update_topology_with_local_metadata`.

Hence, I need to make sure that `update_topology_with_local_metadata` has been called. The `_operation_mode` is set to NORMAL shortly after that function is called. I wanted to avoid adding a `bool` flag just for that, but after all I'll go with that (unless there is some other, better way to do it that I do not know about).

Reply:
I went with a boolean flag for now that is set to true after `update_topology_with_local_metadata` completes successfully.

Reply:
If it something that is expected to happen from time to time it should not be an exception but a return value. Why not return raft_topology_cmd_result::command_status::failure here? Also I do not understand why do we need this check at all. The guard that the coordinator holds while issuing the barrier suppose to handle that. If some node changes its feature's set during the barrier execution writing of enabled features will fail and it will have to be re-calculated.